src/connectors/tls_connector.rs

```diff
@@ -58,30 +72,35 @@ impl<C> TlsConnector<C> {
 
     #[cfg(not(feature = "native-tls"))]
     #[inline]
-    pub fn new_with_tls_default(inner_connector: C) -> Self {
+    pub fn new_with_tls_default(inner_connector: C, alpn: Option<Vec<Vec<u8>>>) -> Self {
         let mut root_store = rustls::RootCertStore::empty();
-        root_store.add_trust_anchors(webpki_roots::TLS_SERVER_ROOTS.iter().map(|ta| {
-            rustls::OwnedTrustAnchor::from_subject_spki_name_constraints(
-                ta.subject,
-                ta.spki,
-                ta.name_constraints,
-            )
-        }));
-
-        let cfg = rustls::ClientConfig::builder()
-            .with_safe_defaults()
+        root_store.extend(webpki_roots::TLS_SERVER_ROOTS.iter().cloned());
+
+        let mut cfg = rustls::ClientConfig::builder()
             .with_root_certificates(root_store)
             .with_no_client_auth();
+
+        // Set ALPN from client side
+        if let Some(alpn) = alpn {
+            cfg.alpn_protocols = alpn;
+        }
+
         TlsConnector::new(inner_connector, cfg.into())
     }
 
     #[cfg(feature = "native-tls")]
     #[inline]
-    pub fn new_with_tls_default(inner_connector: C) -> Self {
-        TlsConnector::new(
-            inner_connector,
-            native_tls::TlsConnector::builder().build().unwrap().into(),
-        )
+    pub fn new_with_tls_default(inner_connector: C, alpn: Option<Vec<Vec<u8>>>) -> Self {
+        let mut tls_connector = native_tls::TlsConnector::builder();
+        if let Some(alpn) = alpn {
+            // Convert alpn to &[&str]
+            let alpns = alpn
+                .iter()
+                .map(|a| std::str::from_utf8(a).unwrap())
```

Comment:
native-tls uses strs as alpn, rustls uses &[u8], so I think it's better to use strs in our function. Convert bytes to str may fail and cause panic or hide problems.

Reply:
Makes sense. Changed to &str